pkg/migrate/migrate.go

```diff
@@ -134,6 +192,82 @@ func Migrate(ctx context.Context, w *log.Logger, clientset k8sclient.Interface,
 	return nil
 }
 
+// NewPVMigrator returns a PV migration context
+func NewPVMigrator(cfg *rest.Config, log *log.Logger, srcSC, dstSC string, podReadyTimeout int) (*PVMigrator, error) {
+	k8scli, err := k8sclient.NewForConfig(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create kubernetes client: %w", err)
+	}
+
+	if srcSC == "" {
+		return nil, fmt.Errorf("empty source storage class")
+	}
+	if dstSC == "" {
+		return nil, fmt.Errorf("empty destination storage class")
+	}
+	if log == nil {
+		return nil, fmt.Errorf("no logger provided")
+	}
+
+	podReadinessTimeout := time.Duration(podReadyTimeout)
+	if podReadinessTimeout == 0 {
+		// default
+		podReadinessTimeout = 60 * time.Second
+	}
+
+	return &PVMigrator{
+		ctx:             context.Background(),
+		log:             log,
+		k8scli:          k8scli,
+		srcSc:           srcSC,
+		dstSc:           dstSC,
+		deletePVTimeout: 5 * time.Minute,
+		podReadyTimeout: podReadinessTimeout * time.Second,
+	}, nil
+}
+
+// PrintPVAccessModeErrors prints and formats the volume access mode errors in pvcErrors
+func PrintPVAccessModeErrors(pvcErrors map[string]map[string]PVCError) {
+	tw := tabwriter.NewWriter(os.Stdout, 0, 8, 8, '\t', 0)
+	fmt.Fprintf(tw, "The following persistent volume claims cannot be migrated:\n\n")
+	fmt.Fprintln(tw, "NAMESPACE\tPVC\tSOURCE\tREASON\tMESSAGE")
+	fmt.Fprintf(tw, "---------\t---\t------\t------\t-------\n")
+	for ns, pvcErrs := range pvcErrors {
+		for pvc, pvcErr := range pvcErrs {
+			fmt.Fprintf(tw, "%s\t%s\t%s\t%s\t%s\n", ns, pvc, pvcErr.from, pvcErr.reason, pvcErr.message)
+		}
+	}
+	tw.Flush()
+}
+
+// ValidateVolumeAccessModes checks whether the provided persistent volumes support the access modes
+// of the destination storage class.
+// returns a map of pvc errors indexed by namespace
+func (p *PVMigrator) ValidateVolumeAccessModes(pvs map[string]corev1.PersistentVolume) (map[string]map[string]PVCError, error) {
+	validationErrors := make(map[string]map[string]PVCError)
+
+	if _, err := p.k8scli.StorageV1().StorageClasses().Get(p.ctx, p.dstSc, metav1.GetOptions{}); err != nil {
+		return nil, fmt.Errorf("failed to get destination storage class %s: %w", p.dstSc, err)
+	}
+
+	pvcs, err := kurlutils.PVCSForPVs(p.ctx, p.k8scli, pvs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get pv to pvc mapping: %w", err)
+	}
+
+	for pv, pvc := range pvcs {
+		v, err := p.checkVolumeAccessModes(pvc)
+		if err != nil {
+			p.log.Printf("failed to check volume access mode for claim %s (%s): %s", pvc.Name, pv, err)
+			continue
+		}
+		if v != (PVCError{}) { // test for empty struct
```

Comment:
can this return *PVCError?

Reply:
Yup. This empty struct predicate looks very cryptic.